marimo/_runtime/copy.py

```diff
@@ -0,0 +1,174 @@
+# Copyright 2024 Marimo. All rights reserved.
+from __future__ import annotations
+
+import inspect
+import sys
+import weakref
+from copy import copy
+from typing import (
+    Any,
+    Callable,
+    Generic,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+T = TypeVar("T")
+if sys.version_info < (3, 9):
+    # Future does not seem to work for this in CI.
+    Ref = Union["weakref.ReferenceType[T]", Callable[[], T]]
+else:
+    Ref = Union[weakref.ReferenceType[T], Callable[[], T]]
+
+
+class CloneError(Exception):
+    """Thrown when strict execution fail to deep copy or clone."""
+
+
+class RoError(Exception):
+    """Thrown when attempting to modify a read-only object."""
+
+
+class _Copy(Generic[T]):
+    """Base wrapper class for strict execution."""
+
+    __ref__: Ref[T]
+
+
+class ZeroCopy(_Copy[T]):
+    """Wrapper class for strict execution (stops deepcopy)."""
+
+
+class ShallowCopy(_Copy[T]):
+    """Wrapper class for strict execution (does copy over deepcopy)."""
+
+
+def _ro_fail() -> None:
+    raise RoError(
+        "Weakly copied objects are directly read only."
+        " Modification is not encouraged, but is"
+        " possible by utilizing the marimo.unwrap_copy() function."
+    )
+
+
+def shadow_wrap(ref_cls: Type[_Copy[T]], base: T) -> T:
+    """
+    Wraps the base object by copying all attributes over to slots, and the then
+    restricting write access to the object attributes / items directly. This is
+    very agrressive and makes the wrapped object difficult to tell apart from
+    the base object without inspect.
+
+    However, there are some limitations, e.g. some operations of the wrapped
+    object may not be associative if they are not defined for both the left
+    and right operations. For instance:
+
+    >>> a = shadow_wrap(ZeroCopy, [1, 2, 3])
+    >>> b = [4, 5, 6]
+    >>> a + b
+    [1, 2, 3, 4, 5, 6]
```

Comment:
Does that mean a zero-copied NumPy array could fail if involved in a binary operation with a vanillia NumPy array?

Reply:
So in numpy

'B + A' works but not 'A + B'